Introduce scopes during tracing

lantiga opened the pull request:
This PR introduces scopes (or namespaces) in order to group operations in the tracing IR, e.g.

```python
x = Variable(torch.Tensor([0.4]), requires_grad=True)
y = Variable(torch.Tensor([0.7]), requires_grad=True)

def doit(x, y):
    tracing_state = torch.jit.get_tracing_state(x,y)
    if tracing_state:
        tracing_state.push_scope('Foo')
    z = Variable(torch.Tensor([0.7]), requires_grad=True)
    out = torch.sigmoid(torch.tanh(x * (y + z)))
    if tracing_state:
        tracing_state.pop_scope()
    return out

traced, _ = torch.jit.trace(doit, (x, y))
g = torch._C._jit_get_graph(traced)
print(g)
```

outputs

```
graph(%1 : Float(1)
      %2 : Float(1)) {
  %3 : Float(1) = Constant[value=<Tensor>](), uses = [%4.i1], scope: Foo;
  %5 : Float(1) = ^Add(False)(%2, %3), uses = [[%6.i1]], scope: Foo;
  %7 : Float(1) = ^Mul()(%1, %5), uses = [[%8.i0]], scope: Foo;
  %9 : Float(1) = ^Tanh()(%7), uses = [[%10.i0]], scope: Foo;
  %11 : Float(1) = ^Sigmoid()(%9), uses = [[%0.i0]], scope: Foo;
  return (%11);
}
```

Scopes work like a stack: they can be pushed and popped manually (as in the example above). Modules automatically push a scope during `__call__` and pops it before returning. The scope is named as `className$id`, where id is the value of the Python `id` function:

```python
class Net(nn.Module):

        def __init__(self):
            super(Net, self).__init__()
            self.layer1 = nn.Sequential(nn.Linear(2,2), nn.ReLU())

        def forward(self, x):
            return self.layer1(x)

    net = Net()

    t = Variable(torch.ones(2), requires_grad=True)

    traced, _ = torch.jit.trace(net, (t, ))
    g = torch._C._jit_get_graph(traced)
    print(g)
```

outputs

```
graph(%1 : Float(2)
      %2 : Float(2, 2)
      %3 : Float(2)) {
  %5 : Float(2!, 2!) = ^Transpose(0, 1)(%2), uses = [[%10.i2]], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %7 : Float(1, 2), %8 : Handle = ^Unsqueeze(0)(%1), uses = [[%10.i1], []], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %9 : Float(1, 2) = Constant[value=<Tensor>](), uses = [%10.i0], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %11 : Float(1, 2), %12 : Handle = ^Addmm(0, 1, True)(%9, %7, %5), uses = [[%13.i0], []], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %14 : Float(2), %15 : Handle = ^Squeeze(0, True)(%11), uses = [[%16.i0], []], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %17 : Float(2) = ^Add(True)(%14, %3), uses = [[%18.i0]], scope: Net$4569918736.Sequential$4569919312.Linear$4569919120;
  %19 : Float(2), %20 : Handle = ^Threshold(0, 0, False)(%17), uses = [[%0.i0], []], scope: Net$4569918736.Sequential$4569919312.ReLU$4569919184;
  return (%19);
}
``` 

*Tests fail at the moment* because the expected output of traces differs, as I added a `scope` description. I'm not sure it belongs there, at the moment it's handy for debugging purposes. If we decide to keep them this way I'll update the expected output.

Under the hood, scope names are implemented using interned strings.

/cc @ezyang @fmassa

lantiga committed:
Add scopes to traces

ezyang commented on torch/nn/modules/module.py:
```diff
@@ -276,6 +293,8 @@ def __call__(self, *input, **kwargs):
                     functools.update_wrapper(wrapper, hook)
                     grad_fn.register_hook(wrapper)
+        if tracing_state:
+            torch._tracing_state.pop_scope()
```
Probably best to try-finally this so that we restore scopes when we pop

lantiga replied:
BTW there is a mistake on my side here (I fixed it on my end but ended up not pushing the fix). It should read
```
if tracing_state:
    tracing_state.pop_scope()
```

Plus try-finally sounds good.

ezyang commented on torch/nn/modules/module.py:
```diff
@@ -20,6 +20,20 @@ def _addindent(s_, numSpaces):
     return s
 
 
+def _first_var(input, kwargs):
```
I'm not sure why this is necessary / why it works at all, because from where I'm sitting `getTracingState` takes a list of variables: `std::shared_ptr<TracingState> getTracingState(const variable_list& vars)`. Shouldn't we instead be using something like flatten rather than `_first_var` here?

lantiga replied:
As for the signature, in `torch/jit/__init__.py` I wrapped `torch._C.get_tracing_state`, so that it's accessible as `torch.jit.get_tracing_state`. I chose to expand the `*args` from the former in the signature of the latter, since this is an operation that the user would have to do frequently when pushing and popping.

As to `flatten` vs `_first_var`: `get_tracing_state` needs at least one variable, so I just stop at the first I find. Is there an advantage with using more than one? Maybe we should check that the variable we encounter is traced at all?

ezyang replied:
Oh, I missed the `*args`!

The advantage to doing more than one is (1): some variables may not be traced (this occurs if someone creates a Variable during the course of tracing), but as long as *one* variable is traced, the expression participates in the trace, (2) we get to do more sanity checking in this case (because we can check that the vars are consistent.)

lantiga replied:
Got it, changing to flatten

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -152,6 +157,23 @@ struct Node : public Attributes<Node> {
+    if (scope_.empty()) {
+      return scope_name;
+    }
+    scope_name = std::accumulate(
```
Apologies, but looking at this part of the patch I am now not sure about the wisdom of making these interned strings. The original motivation for interning was to make it easier to test for equality on scopes, but it looks like this isn't actually being used ever, so maybe we shouldn't intern here. I wonder if @zdevito has thoughts.

lantiga replied:
I was thinking about this. The advantage right now is that it might be lighter if you have a lot of nodes, but you're right, we're not matching at the moment.
Could we ever expose interned strings to python without converting to strings?

I'm fine reverting to strings anyway.

ezyang replied:
I'd wait and see if anyone has opinions. If not, let's perhaps not bother changing it; we're allocating a string per scope and there should only be finitely many of these.

apaszke replied:
Interning makes sense, because without it every Node would keep a copy of names of all parent scopes, which is quite a lot. However, if someone uses varying scope names this can really pollute the interned pool. I think it would be nicer to implement scoping mechanism in a trie-like approach, where scopes build up a tree (with `shared_ptr`s to parent nodes) with names at each level, and `Node`s only point to their node in the scope tree.

ezyang commented:
This looks pretty good to me.

lantiga commented:
Thanks for the review @ezyang

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -254,28 +268,35 @@ def register_forward_hook(self, hook):
+        tracing_state = torch.jit.get_tracing_state(_first_var(input, kwargs))
+        try:
+            if tracing_state:
+                tracing_state.push_scope('%s$%d' % (self.__class__.__name__, id(self)))
```
Instead of naming the scope with the name of the `Module` class, I think it would be nicer if we assign the name of instance.
One way of achieving this would be to assign a `self._name` attribute into the class whenever we add a module. This way, the names of the scopes would be the same as the names that you give to the modules, and the same that you obtain when printing the network. Also, there would be no need of adding the `id(self)` to the scope, as the names would be unique.

What do you think?

ezyang replied:
Yes, it would be good if the computed name here corresponds with however we identify multiple copies of a module inside a parent module.

lantiga replied:
I thought about it, it would be good to assign a `_name` property to a specific instance of a module at the constructor level.
If the user doesn't provide the instance name, though, we have two options:
- use the `id`, quick but not that readable
- assign the name using a self-incrementing counter, as a class-level member variable; the gotcha here is what happens when one loads a serialized model
What do you think?

fmassa replied:
There is always a name, because we store the modules in an `OrderedDict`, so that should be fine I think?
Let me know if this isn't clear, I can try writing a snippet.

lantiga replied:
Ah-ha, you're right!

fmassa replied:
One possibility for capturing the name of a module is to add it to the module in both [`__setattr__`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L323-L328) and [`add_module`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L111-L121). 
So you'd do something like (in `add_module`):
```python
module._name = name
self._modules[name] = module
```
Maybe it's not necessary to add a new attribute to the module, and we can use the names in `self._modules` directly.

On an unrelated note, it's a pity that `__setattr__` doesn't call `add_module` internally I think, but well, that's for another discussion.

lantiga replied:
Done!

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -254,28 +268,35 @@ def register_forward_hook(self, hook):
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        tracing_state = torch.jit.get_tracing_state(_first_var(input, kwargs))
```
nit: I think it would be nicer to have a context manager for handling the scopes, which would take care of the `try`/`finally` for you.

lantiga replied:
I like this, we should make it (optionally?) available to users too

fmassa replied:
Yes, I think it would be good to make this available at some point, once we have converged the interface nicely.

lantiga replied:
Done

fmassa commented:
I think this looks good, thanks @lantiga! I have some small comments that are mostly details, let me know what you think.

lantiga commented:
I've just pushed a new commit:
* args to `__call__` in module are flattened (instead of `_first_var`)
* I introduced a `torch.jit.scope` context manager

So, what was
```
x = Variable(torch.Tensor([0.4]), requires_grad=True)
y = Variable(torch.Tensor([0.7]), requires_grad=True)

def doit(x, y):
    tracing_state = torch.jit.get_tracing_state(x,y)
    if tracing_state:
        tracing_state.push_scope('Foo')
    z = Variable(torch.Tensor([0.7]), requires_grad=True)
    out = torch.sigmoid(torch.tanh(x * (y + z)))
    if tracing_state:
        tracing_state.pop_scope()
    return out

traced, _ = torch.jit.trace(doit, (x, y))
g = torch._C._jit_get_graph(traced)
print(g)
```

becomes

```
x = Variable(torch.Tensor([0.4]), requires_grad=True)
y = Variable(torch.Tensor([0.7]), requires_grad=True)

def doit(x, y):
    with torch.jit.scope('Foo', (x, y)):
        z = Variable(torch.Tensor([0.7]), requires_grad=True)
        out = torch.sigmoid(torch.tanh(x * (y + z)))
    return out

traced, _ = torch.jit.trace(doit, (x, y))
g = torch._C._jit_get_graph(traced)
print(g)
```

fmassa commented on torch/jit/__init__.py:
```diff
@@ -28,6 +28,25 @@ def __repr__(self):
+        tracing_state = torch._C.get_tracing_state(vars)
+    if tracing_state:
+        tracing_state.push_scope(scope_name)
+    yield
```
I think it might be good to wrap this in a `try`/`finally`, so that if errors occur we still properly clean up the `tracing_state`.
Maybe something like
```python
try:
    yield
finally:
    if tracing_state:
        tracing_state.pop_scope()
```

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -20,6 +20,20 @@ def _addindent(s_, numSpaces):
     return s
 
 
+def _flatten(*args):
```
nit: I wonder if @ezyang was talking about [this `_flatten`](https://github.com/pytorch/pytorch/blob/master/torch/jit/__init__.py#L540-L543)

lantiga commented:
Thanks @fmassa BTW!

lantiga commented:
Here's the new output (each module appears as `className$name`):
```
graph(%1 : Float(2)
      %2 : Float(2, 2)
      %3 : Float(2)) {
  %5 : Float(2!, 2!) = ^Transpose(0, 1)(%2), uses = [[%10.i2]], scope: Net$inner.Sequential$layer1.Linear$0;
  %7 : Float(1, 2), %8 : Handle = ^Unsqueeze(0)(%1), uses = [[%10.i1], []], scope: Net$inner.Sequential$layer1.Linear$0;
  %9 : Float(1, 2) = Constant[value= 0.5175 -0.6471 [ CPUFloatTensor{1,2} ]](), uses = [%10.i0], scope: Net$inner.Sequential$layer1.Linear$0;
  %11 : Float(1, 2), %12 : Handle = ^Addmm(0, 1, True)(%9, %7, %5), uses = [[%13.i0], []], scope: Net$inner.Sequential$layer1.Linear$0;
  %14 : Float(2), %15 : Handle = ^Squeeze(0, True)(%11), uses = [[%16.i0], []], scope: Net$inner.Sequential$layer1.Linear$0;
  %17 : Float(2) = ^Add(True)(%14, %3), uses = [[%18.i0]], scope: Net$inner.Sequential$layer1.Linear$0;
  %19 : Float(2), %20 : Handle = ^Threshold(0, 0, False)(%17), uses = [[%0.i0], []], scope: Net$inner.Sequential$layer1.ReLU$1;
  return (%19);
}
```

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -56,6 +70,7 @@ def __init__(self):
         self._forward_pre_hooks = OrderedDict()
         self._modules = OrderedDict()
         self.training = True
+        self._name = ''
```
Because we have added a new field to the module, I wonder if we shouldn't adapt the `__getstate__` to recreate the missing fields for backward-compatibility reasons. @ezyang what do you think?

ezyang replied:
I don't have very much experience with Python's pickler, so I can't say for certain, but it is definitely better for people's old pickles to keep unpickling after this change. I don't know what is necessary to actually make this happen.

lantiga replied:
Unpickling will still work, but the `name_` attribute won't be present, so the code will fail. I'm adding a
```
    def __setstate__(self, d):
        self.__dict__.update(d)
        self.__dict__.setdefault('name_', '')
```
to Module so that the field will still exist after unpickling an old pickle. I don't think we need a `__getstate__`, but I may be missing something.

lantiga replied:
Also: we could set the default name to something unique. Do you have ideas?

fmassa replied:
Yeah, I think we just need to modify the deserialization part (`__setstate__`, I think I confused the names :-) ).
I think the only situation where the `_name` won't be assigned to a name is in the main model class, all others will have names, right? In this case, I think we could name it something like `model`?

lantiga replied:
Thanks @fmassa , I have modified `__setstate__` in the last commit.
I was especially referring to unpickling of old pickles in `__setstate__`. Right now it will just make sure that `_name` exists, but it will set it to the empty string. This will lead to incorrect grouping of nodes for older models, if one were to, say, display a pre-trained network.

Using `id(self)` in these cases might be ok, although it's not stable in subsequent unpicklings.
Alternatively we have to assign a progressive number to each unpicked object of a given class, although this is a bit more involved.

What do you think?

fmassa replied:
Ah, I misunderstood what you meant.
It is possible to traverse the structure the modules and assign `_name` to each of the modules, so that the new behaviour can be found in old serialized models as well.
Here is an approximate snippet:
```python
for name, module in self.named_modules():
    module._name = name
```
What do you think?

fmassa replied:
(note: `named_modules` returns each module only once, so the namings might not be consistent because the name of a module would be the last assigned one during model creation, but the first one during deserialization)

fmassa replied:
(we should also note that such kind of problems would only happen if the user serialized the whole `Module` object, and not if he serialized the `state_dict`, which is the preferred way to do it)

lantiga replied:
What about we use `named_children` in `__setstate__`? Since all instances of modules will be unpickled and children will have to be unpickled first for the unpickling to succeed, at the end all modules will be named (except the top-most, which we don't really care about)

lantiga replied:
`__setstate__` would become:
```
    def __setstate__(self, state):
        self.__dict__.update(state)
        if '_name' not in self.__dict__:
            self._name = ''
            for name, module in self.named_children():
                module._name = name
```

fmassa replied:
Looks good to me as well.

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -510,6 +534,27 @@ friend struct Node;
+  }
+
+  Graph* pushScope(const std::string& scope) {
+    if (scope.empty()) {
```
This seems like an odd thing to do. If you accidentally passed an empty string to the decorator, won't you ignore the push but then unconditionally perform the pop?

lantiga replied:
Good catch. That check is mostly vestigial, removing it.

fmassa commented on torch/nn/modules/module.py:
```diff
@@ -331,14 +334,16 @@ def remove_from(*dicts):
                                     .format(torch.typename(value), name))
-                value._name = name
+                if value is not None:
+                    value._name = name
```
nit: I just saw that there is a `raise TypeError` just above, `if value is not None`. We can then maybe remove this condition + assignment altogether?

lantiga replied:
Of course, thanks @fmassa

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -254,32 +257,38 @@ def register_forward_hook(self, hook):
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        with torch.jit.scope('%s$%s' % (self.__class__.__name__, self._name),
```
I don't think adding this in here is a good idea. It's a quite large context-manager, implemented using `contextlib` and I'm pretty sure it will be causing us performance problems. It will always be iterating over the input, always does the string interpolation, always enters, always goes to C to check Variable flags, etc. This is definitely too much considering that in 99.999% of invocations all this will be unnecessary. We should come up with a way to make this cheap before this is merged.

Also, I'm not a huge fan of the `$` syntax. Maybe let's just do `ClassName<user specified name>` and `ClassName` if nothing was given. This will let you do e.g. `Conv2d_1` and if someone likes `$` they can still add that to their names.

lantiga replied:
I **really** like the trie idea!

As to the cost: would it be ok to pay for it only if we're tracing? We can also reduce the cost by going back to `push_state` and `pop_state` and keeping the context manager for willing users.

OK for the syntax.

ezyang replied:
Yeah, I think paying only when tracing is fine.

lantiga replied:
I removed the context from `__call__` and introduced a `torch.jit._tracing` module-global variable that is `True` when something is tracing and `False` otherwise. This should bring the cost of scopes while not tracing (which is almost always) to ~zero.

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -254,32 +257,38 @@ def register_forward_hook(self, hook):
+        if '_name' not in self.__dict__:
+            self._name = ''
+            for name, module in self.named_children():
+                module._name = name
```
`__setstate__` will also be called for children, so this unnecessarily makes loading the tree O(n^2) instead of O(n)

lantiga replied:
So, just to make sure I get it right: `__setstate__` will be called once per module, and each module will iterate its _immediate_ children and set their names. Wouldn't this operation only touch each module once, not multiple times?

lantiga replied:
@apaszke I didn't make changes here, but please let me know if I'm getting it wrong.

ezyang replied:
Yes, this looks fine to me.

apaszke replied:
My bad. The complexity is correct, but the code isn't. If you have e.g. 4 levels of nesting like this, the names will end up as:
```
m1            ''
  m2          'm2'
    m3        ''     # NOTE: m2 had _name set, so it didn't iterate
      m4      'm4' 
```

There's still a bit of a problem with models that are DAGs - see my comment below

lantiga replied:
Nice catch! Of course it wouldn't work.
This
```
if '_name' not in self.__dict__:
    self._name = ''
for name, module in self.named_children():
    if '_name' not in module.__dict__:
        module._name = name
```
would fix it, at the cost of checking all children even if `name_` is in the parent.

However, thinking about DAG, it may make no sense to assign a name attribute to a module itself (which is why it wasn't there in the first place 😄, I get it now). 

I'm realizing that the way to get the name for a module (to auto-generate the scope name for the module) is to obtain it from the stack you're talking about in your DAG comment.

apaszke replied:
Yup, the stack is a reliable way to get this, without all the `_name` maintenance headaches.

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -245,7 +245,8 @@ std::ostream& printNode(std::ostream & out, Node * n, std::vector<Node*> * group
   }
-  out << "];\n";
+  out << "], ";
+  out << "scope: " << n->scopeName() << ";\n";
```
can you make printing of the scope optional based on whether it is defined? It's unnecessary if it was not used and this way we won't have to change all expect files as well :)

lantiga replied:
Done

apaszke reviewed (changes requested):
Awesome, good to finally get scopes :) I have a few requests regarding implementation details

lantiga commented:
Hey @apaszke @ezyang , I had a very packed couple of weeks. I resumed this - I've got the trie, will make the last changes and update the PR. Thanks for your patience!

lantiga commented:
Note after last update: scopes are not preserved after a ONNX pass. I'm looking into it.

lantiga commented:
Fixed scopes for the ONNX pass.

apaszke commented:
@lantiga we'll have to wait with merging this PR anyway. Autograd PR heavily disrupted the JIT work, but it was very high priority, so we were ok with this. Right now, we're focused on unbreaking ONNX, but parts of the JIT infra will be different now - e.g. we won't have this initial onnxification pass, only the proper export pass (most functions are in C++ and we already auto-generate code to make them JIT nodes with proper names). Sorry for having you wait 😕

lantiga commented:
No worries, I had to get this out of my system. When you’re ready I’ll be happy to fix any breakages that the various merges might have caused. Thanks

lantiga commented:
Since the C++ refactor has landed, this PR has been rebased on top of current master and updated accordingly.

A few tests fail because current expects do not include scope information. Scopes are now added automatically by `Module` if `tracing` is active, so they then get printed.
We can either update the expects or make printing include scopes through a flag.

/cc @ezyang  @apaszke

ezyang commented on torch/csrc/jit/ir.h:
```diff
@@ -60,6 +61,65 @@ static inline bool operator==(const Use & a, const Use & b) {
+// to the scope that was current when the node was created.
+// The trie never needs to shrink, it only grows until it is disposed
+// of when Graph is deallocated. Hence, pointers to scopes held by nodes
+// will always be valid.
```
It's worth noting that the ownership model as specified in the destructor implies that you cannot share common subtries between tries.

lantiga replied:
Exactly, I couldn't think of a reason why different Graphs might share portions of the trie. I'm writing a note.

ezyang commented:
> A few tests fail because current expects do not include scope information. Scopes are now added automatically by Module if tracing is active, so they then get printed. We can either update the expects or make printing include scopes through a flag.

The scopes we add by default are fairly stable, right? I'd be happy with either solution, but certainly `--accept` ing the new output seems easiest.

ezyang commented:
I think this is good enough to be useful, so we should merge it before it bitrots more, but since @zdevito is owner of the JIT/tracer, he makes the final call.

Having done one more review pass on it, there are two high level things that pop out to me:

* This PR adds another "observation point" for seeing tracing, the global `torch.jit._tracing`. I can see why this is expedient, and I think it is being used in a sound way, but it means that backwards tracing with scopes will not work out of the box. Since the primary use-case for this was ONNX (forwards only), I think it makes sense not to try to also nail the backwards case today; furthermore, I suspect @zdevito may be touching the tracer a lot in the near future, so the mechanics of how to do this correctly may change.

* Scopes were attached as state to the graph. This is consistent with the way that we've handled other similar concerns (e.g., stages), but it is also morally wrong: logically, an IR doesn't have a "scope" in any sense; it's a property of the tracer. But I don't think it would be awful to fix this later, if we ever need it fixed, so it seems OK for now as well.

lantiga commented:
Thank you @ezyang! There was a conflict, I rebased on top of master.

> This PR adds another "observation point" for seeing tracing, the global torch.jit._tracing [...]

I wanted to avoid that, but I saw no other way. Note (this is for others who may approach this PR) that this variable is only used to make scopes inexpensive at the `Module` level when tracing is not active (modules create scopes automatically when they are called). This global is not used elsewhere.

> Scopes were attached as state to the graph. [...]

Indeed I had them attached to the TracingState in a previous iteration. However, a Graph Node will need to hold information on what scope it's pointing to (i.e. what scope was the current one when it was created). Since the lifetime of Nodes is tied to the lifetime of its hosting Graph, it is quite natural that the scopes trie is kept in the Graph.
I mean, I see your point, but keeping the trie in TracingState is more dangerous in my opinion.

lantiga commented:
> The scopes we add by default are fairly stable, right? I'd be happy with either solution, but certainly --accept ing the new output seems easiest.

Last thing: yes, scopes are stable as long as the name of what gets registered in Modules is stable. If the name of a Module variable changes, then the respective expect will be invalid.

I don't have an issue adding a flag to disable the output of scopes, just let me know.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -136,6 +199,9 @@ struct Node : public Attributes<Node> {
   bool hasType() const {
     return type_ != nullptr;
   }
+  virtual std::string name() const {
```
Do we need to add this method? The problem with names is that we already have `uniqueName()` and `debugName()`, and `name()` sounds like sth even more important, but really it returns an empty string for most of our nodes

lantiga replied:
We don't, removing.

apaszke commented on torch/csrc/jit/python_tracer.cpp:
```diff
@@ -32,6 +32,14 @@ void initPythonTracerBindings(PyObject* module_) {
+      s.push_scope(scope_name);
+    })
+    .def("pop_scope", [](TracingState& s) {
+      ASSERT_UNEXPIRED("export");
```
nope, not "export"

lantiga replied:
Whoops, sorry. Fixed.

apaszke commented on torch/jit/__init__.py:
```diff
@@ -235,7 +262,10 @@ def forward(self, *args):
+    global _tracing
+    _tracing = True
     out = f(*args)
+    _tracing = False
```
Note: this is only called from `torch.jit.trace`. `torch.jit.compile` won't use scopes at all in this patch (which might not be such a bad thing - see the comment on optimizations)

apaszke commented on torch/jit/__init__.py:
```diff
@@ -19,6 +19,33 @@
+    if not torch._C._is_tracing(vars):
+        tracing_state = None
+    else:
+        tracing_state = torch._C._get_tracing_state(vars)
```
It seems like you duplicated the check above.

lantiga replied:
Fixed

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -391,6 +408,8 @@ def remove_from(*dicts):
                         "cannot assign module before Module.__init__() call")
                 remove_from(self.__dict__, self._parameters, self._buffers)
+                if value is not None:
+                    value._name = name
```
Treating modules like this is a bit awkward, because we don't enforce the model to form a tree - it's more of a DAG. The same module can be a submodule of multiple parents, and it happens in a few models. A better solution would be to keep a stack of modules, and inside that `slow_forward()` function I mentioned you could look up the topmost module, fine the name of the current module in the outer scope, and use that. This would let modules have multiple names and remove the need for this attribute. And it's tracing so we don't care this much about perf

lantiga replied:
Done

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -320,32 +323,46 @@ def register_forward_hook(self, hook):
-                    grad_fn.register_hook(wrapper)
+        tracing_state = None
+        if torch.jit._tracing:
+            input_vars = tuple(torch.autograd.function._iter_variables(input))
```
This is worrying, because this means that we won't be able to trace models that take arbitrary things as inputs (`_iter_variables` has some assumptions about types - they should be nested tuples, lists and dicts of Variables)

lantiga replied:
I see. I use `input_vars` to obtain `tracing_state` with `torch.jit.get_tracing_state`, which is only interested in Variables, if I got it right.
I don't have other ideas except accessing the tracing state from somewhere else...

apaszke commented on torch/nn/modules/module.py:
```diff
@@ -320,32 +323,46 @@ def register_forward_hook(self, hook):
+                        grad_fn.register_hook(wrapper)
+        finally:
+            if tracing_state:
+                tracing_state.pop_scope()
```
I think that it would be better to do this to reduce overhead (and simplify code):
```python
# do the pre-hooks
if torch.jit._tracing:
    result = self.slow_forward() # set the scope and only then call self.forward(*input, **kwargs) in a try block
else:
    result = self.forward(*input, **kwargs)
# do the post hooks
```

lantiga replied:
Good idea, fixed

apaszke reviewed (commented):
That looks much better, thanks Luca! I left some comments and proposed alternative implementation for module scoping that would be more flexible

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -60,6 +61,67 @@ static inline bool operator==(const Use & a, const Use & b) {
+private:
+  Scope* parent_;
+  Symbol name_;
+  std::unordered_set<const Scope*> children_;
```
So, Graph holds a pointer to a node in the trie (the current node) and another pointer to the root. They are both pointers to Scope. The current node is for sure a member of `children_` of some other node, yet I have to hold it as current.

I'm not sure how I would safely use `unique_ptr` to store pointers to children without having to  `get` the raw pointers out to set them as current. It seems it's still an unsafe thing to do.
Should we use a `shared_ptr` to avoid these troubles?

apaszke replied:
It seems to me that using a `unique_ptr` for the root, and `Scope*` for current scope, while having `unique_ptr`s in the scope trie for children will be safe - current scope is guaranteed to stay alive, because it is present in the subtree rooted in `root_scope_`, so there exists a chain of `unique_ptr`s leading to that `Scope`. I think it's exactly the same ownership model as proposed in this PR.

Calling `.get()` is safe is you can ensure that the pointer you received will be alive as long as you'll be using the result, which is the case here.

lantiga replied:
Got it, thanks!

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -60,6 +61,67 @@ static inline bool operator==(const Use & a, const Use & b) {
+  }
+  Scope* pop() {
+    if (parent_ == NULL) {
+      return this;
```
Done

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -60,6 +61,67 @@ static inline bool operator==(const Use & a, const Use & b) {
+      current = current->parent_;
+    }
+    return out;
+  }
```
Yeah, I was reversing before concatenating in Graph. Now I chose to just output the concatenated string directly, since I'm not using the flat list of scope nodes to Root for anything else.

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -156,6 +222,29 @@ struct Node : public Attributes<Node> {
+                    [] (std::string a, Symbol b) {
+                    return a + '.' + std::string(symbolToString(b));
+                    });
+    return scope_name;
```
Done :-)

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -533,6 +618,9 @@ friend struct Node;
   size_t new_node_stage_;
 
+  std::shared_ptr<Scope> scope_root_;
+  Scope * scope_;
```
Fixed

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -543,6 +631,15 @@ friend struct Node;
+  , scope_(scope_root_.get())
+  , output_(initOutput(create(kReturn))) {}
+
+  Graph(std::shared_ptr<Scope>& scope_root)
```
Not after we move scopes out of Graph.

apaszke replied:
But even if they were to be part of the graph when would you create a graph that has a custom root?

lantiga replied:
I was looking at `createFusionGroup` (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir.h#L679) and `ToONNX` (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/onnx.cpp#L36), but I might be wrong here.

lantiga replied:
Moved scope to `TracingState`

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -596,6 +693,32 @@ friend struct Node;
+  
+  // TODO: Scope is most accurately a property of tracing, rather than
+  // of the IR (which doesn't really know or care about the scopes.)
+  // Move this to TracingState eventually.
```
Moving to `TracingState` has the following gotcha: `Node`s hold a pointer to nodes of the `Scope` trie. If we assume that `Node`s always belong to the same `Graph` (or a graph cloned from it, so we can pass the `shared_ptr` to `Scope`s along), storing the `Scope` trie in the `Graph` ensures that the pointers to scopes in `Node`s will always be valid. Both the lifetime of `Nodes` and `Scope` depend on the lifetime of `Graph`.

Can we have similar guarantees if we move `Scope` to `TracingState`?

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -533,6 +618,9 @@ friend struct Node;
 
   size_t new_node_stage_;
 
+  std::shared_ptr<Scope> scope_root_;
```
Made a `unique_ptr` in `TracingState`.

lantiga commented:
@apaszke please hold on on reviewing, I need to rethink where I’m stashing the module stack...

apaszke commented:
Why not in `torch/jit/__init__.py` right next to `_tracing` flag?

lantiga commented:
Indeed that's the first spot where I placed it. But I didn't like that jit had know about modules.
My primary concern is that it's effectively a global stack.

Wouldn't it be better to still define it in `module.py`, but attach it to the `tracing_state` object? This way each tracing state would have its own independent stack. But of course we don't want tracing state know about modules.

So I was thinking that maybe we would have a field in tracing state where one can store extra info to be preserved during tracing. What do you think?

apaszke commented:
Right, that makes a lot of sense. It needs to be attached to the tracing state. `module.py` seems reasonable.

lantiga commented:
Done, I activated dynamic attributes on TracingState in pybind and I'm managing a stack per trace now.

lantiga commented:
Ok, I'm done with the changes.

lantiga committed:
Introduce scope API in Value

lantiga committed:
Align with change in trace string output

lantiga commented:
@pytorchbot test this please

lantiga committed:
Expand context manager args

lantiga committed:
Update scopes expect to new JIT output

lantiga commented:
@apaszke is there anything else that needs to be taken care of? Just checking

apaszke commented:
@lantiga sorry I didn't have time to look into it yet. I'll try to do it soon

lantiga commented:
Thank you @apaszke! No worries, take your time

apaszke reviewed (commented):
This looks really good, except for the Scope ownership scheme (this cycle I mentioned in the comment). Once this is fixed, it should be good to go.

lantiga commented on torch/csrc/jit/tracer_state.h:
```diff
@@ -59,6 +61,9 @@ struct TracingState : public std::enable_shared_from_this<TracingState> {
   std::shared_ptr<Graph> graph;
   bool active;
 
+  std::unique_ptr<Scope> scope_root;
```
I certainly misinterpreted. 
Can you clarify a few things for me:

- In the previous iteration the scopes trie was owned by the root of the graph, as you suggest in 1. We can go back to that design and use a shared pointer for those cases in which a graph is cloned and nodes need to keep pointing to the same scope trie nodes.

- Having pointers in the direction `scopes` -> `ir` requires a greater change, but if we like it more we can go for it. In this case we can't tell a node to store the pointer to the trie node that was current when the node was created. Rather, the scopes trie would have to store pointers to nodes in a set (or vector) at each trie node. This would make Graph scope-free, what do you think about this option?

Thanks for the awesome review BTW

apaszke replied:
I think the second option is nicer overall, but I don't want to stall this PR for any longer, so let's go with 1. for now. We can fix it in a later PR. Is that ok with you?

lantiga replied:
Yep, fine with me. Once this is merged I can work at the refactoring, but probably it's also good to see how #3906 impacts the design

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -69,6 +69,65 @@ struct SourceLocation {
+private:
+  Scope* parent_;
+  Symbol name_;
+  std::unordered_set<std::unique_ptr<Scope> > children_;
```
The only reason is that you could theoretically `push` a scope that is already there, with the effect that you make a previous scope become current again. With a vector, we simply have to check before we insert to avoid duplication - this is done infrequently and on a small vector, so it makes perfect sense.

apaszke replied:
Ohh, right. let's keep it as a set then!

lantiga commented on torch/csrc/jit/ir.h:
```diff
@@ -69,6 +69,65 @@ struct SourceLocation {
+    children_.insert(std::unique_ptr<Scope>(newScope));
+    return newScope;
+  }
+  Scope* pop() {
```
Done

lantiga commented on torch/nn/modules/module.py:
```diff
@@ -319,10 +319,42 @@ def register_forward_hook(self, hook):
         self._forward_hooks[handle.id] = hook
         return handle
 
+    def tracing_name(self, tracing_state):
```
Done

lantiga commented on torch/nn/modules/module.py:
```diff
@@ -319,10 +319,42 @@ def register_forward_hook(self, hook):
+            return None
+        module = tracing_state._traced_module_stack[-1]
+        for name, child in module.named_children():
+            if child == self:
```
Done

lantiga commented on torch/nn/modules/module.py:
```diff
@@ -319,10 +319,42 @@ def register_forward_hook(self, hook):
+                return name
+        return None
+
+    def slow_forward(self, *input, **kwargs):
```
Done

lantiga committed:
Move ownership of Scope back to Graph. Minor fixes.

lantiga commented:
@apaszke Fixes are in, Scopes are back into Graph.

lantiga committed:
Updated comment to reflect ownership change

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -70,6 +70,65 @@ struct SourceLocation {
+  Scope* push(Symbol name) {
+    Scope* newScope = new Scope(this, name);
+    children_.insert(std::unique_ptr<Scope>(newScope));
+    return newScope;
```
Using set is fine, but as far as I understand the main purpose is to make it a true `trie`, but this is not really what happens in here - you will return the new unbound scope even if the insertion failed, which in turn will cause the pointer to be dangling, because the `unique_ptr` will be destructed immediately.

Now I'm wondering if we really want to keep it as a true trie... Maybe it just makes sense to have a new scope for each invocation of a module on a particular level (so you can actually see a sequence of scopes - if you call A B A, then you will see A B A in the scope tree).

 Maybe let's really use a vector and skip deduplication.

lantiga replied:
Uh oh, sorry about this one. In fact, I was thinking about it the wrong way when I said you would re-enter a scope.
The insert would always succeed here, since the scope is always a new scope and the set will by construction not contain it. Which is equivalent to using a vector and appending to it, so I'm just going to use a vector.

Deduping based on names only makes sense if we'll ever compare the `scope_` pointers stored in Node to check if two nodes have the same scope. I'm going without, let me know about this.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -567,6 +646,15 @@ friend struct Value;
   , new_node_stage_(0)
+  , scope_root_(std::make_shared<Scope>())
+  , current_scope_(scope_root_.get())
+  , output_(initOutput(create(kReturn, 0))), input_(create(kParam, 0)) {}
```
nit: use constructor delegation here

lantiga replied:
Done

apaszke reviewed (commented):
Looks good, but there's this one problem that can lead to segfaults. Should be ready once this is fixed.

lantiga committed:
Use vector as children container in scope trie

lantiga commented:
@apaszke Done. I'm using `vector` now (see discussion inline). I only quote this line:

> Deduping based on names only makes sense if we'll ever compare the scope_ pointers stored in Node to check if two nodes have the same scope. I'm going without, let me know about this.

as I'd like to make sure with you about deduping.

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -93,7 +93,7 @@ struct Scope {
   Scope* push(Symbol name) {
     Scope* newScope = new Scope(this, name);
-    children_.insert(std::unique_ptr<Scope>(newScope));
+    children_.push_back(std::unique_ptr<Scope>(newScope));
```
In general the problem with this line is that it can leak memory if `push_back` fails (e.g. because an allocation fails). It would be better to do this:
```cpp
children_.push_back(std::unique_ptr<Scope>(new Scope(this, name)));
return children_.back().get();
```

apaszke reviewed (approved):
Looks good, but it conflicts with something now 😕 I also noticed that it could leak memory in case of an error so it would be nice to fix that as well

lantiga committed:
Avoid segfault in case of memory allocation error

lantiga commented:
@apaszke fixed!

apaszke commented:
Thanks Luca!

hhsecond commented:
@lantiga @apaszke  I am writing a visualizer for PyTorch using the tracer and scope in tracer is extremely useful for grouping the nodes. I understand that each node has a pointer to its scope but I don't think we have access to that from python. Wouldn't it be nice to get the scope name in python probably as an attribute of node objects like ``node.scope()`` or something?

lantiga commented:
@hhsecond adding the scope name from python is not a problem. You will get a string `Foo/Bar` or `Foo/Bar[baz]`, where `Foo` and `Bar` are scope names. The second form will be generated automatically by modules, one scope name per module: in this case `baz` is the name of the variable containing the module `Bar` in the parent module `Foo`
```
class Foo(nn.Module):
    def __init__(self):
        [...]
        self.baz = Bar()
```

Feel free to open an issue for exposing scopeName to python and mention me.

hhsecond commented:
@lantiga Perfect! Thanks a ton

lanpa commented:
@lantiga Will scope information be saved in onnx too? Thanks

lantiga commented:
I don't know about what the specific plans are /cc @ezyang and @bddppq
